File: C1.UWP.FlexChart/CS/FinancialChartExplorer/Ext/DateTimeExtends.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FinancialChartExplorer
{
    public static class DateTimeExtends
    {

        public static double ToOADate(this System.DateTime date)
        {
            return TicksToOADate(date.Ticks);
        }

        public static DateTime FromOADate(this System.DateTime date, double val)
        {
            return new DateTime(DoubleDateToTicks(val), DateTimeKind.Unspecified);
        }


        private static double TicksToOADate(long value)
        {
            if (value == 0L)
            {
                return 0.0;
            }
            if (value < 864000000000L)
            {
                value += 599264352000000000L;
            }
            if (value < 31241376000000000L)
            {
                throw new OverflowException("Arg_OleAutDateInvalid");
            }
            long num = (value - 599264352000000000L) / 10000L;
            if (num < 0L)
            {
                long num2 = num % 86400000L;
                if (num2 != 0L)
                {
                    num -= (86400000L + num2) * 2L;
                }
            }
            return (double)num / 86400000.0;
        }

        internal static long DoubleDateToTicks(double value)
        {
            if (value >= 2958466.0 || value <= -657435.0)
            {
                throw new ArgumentException("Arg_OleAutDateInvalid");
            }
            long num = (long)(value * 86400000.0 + ((value >= 0.0) ? 0.5 : -0.5));
            if (num < 0L)
            {
                num -= num % 86400000L * 2L;
            }
            num += 59926435200000L;
            if (num < 0L || num >= 315537897600000L)
            {
                throw new ArgumentException("Arg_OleAutDateScale");
            }
            return num * 10000L;
        }
    }
}

File: C1.UWP.FlexChart/VB/FinancialChartExplorer/Ext/DateTimeExtends.vb

Public Module DateTimeExtends

    <Extension()>
    Public Function ToOADate([date] As DateTime) As Double
        Return TicksToOADate([date].Ticks)
    End Function

    <Extension()>
    Public Function FromOADate([date] As DateTime, val As Double) As DateTime
        Return New DateTime(DoubleDateToTicks(val), DateTimeKind.Unspecified)
    End Function


    Private Function TicksToOADate(value As Long) As Double
        If value = 0L Then
            Return 0.0
        End If
        If value < 864000000000L Then
            value += 599264352000000000L
        End If
        If value < 31241376000000000L Then
            Throw New OverflowException("Arg_OleAutDateInvalid")
        End If
        Dim num As Long = (value - 599264352000000000L) / 10000L
        If num < 0L Then
            Dim num2 As Long = num Mod 86400000L
            If num2 <> 0L Then
                num -= (86400000L + num2) * 2L
            End If
        End If
        Return CDbl(num) / 86400000.0
    End Function

    Friend Function DoubleDateToTicks(value As Double) As Long
        If value >= 2958466.0 OrElse value <= -657435.0 Then
            Throw New ArgumentException("Arg_OleAutDateInvalid")
        End If
        Dim num As Long = CLng(value * 86400000.0 + (If((value >= 0.0), 0.5, -0.5)))
        If num < 0L Then
            num -= (num Mod 86400000L) * 2L
        End If
        num += 59926435200000L
        If num < 0L OrElse num >= 315537897600000L Then
            Throw New ArgumentException("Arg_OleAutDateScale")
        End If
        Return num * 10000L
    End Function
End Module
